proxy/plugin/filter_by_url_regex.py

```diff
@@ -0,0 +1,135 @@
+    :license: BSD, see LICENSE for more details.
+"""
+
+import logging
+
+from typing import Optional
+
+from ..http.exception import HttpRequestRejected
+from ..http.parser import HttpParser
+from ..http.codes import httpStatusCodes
+from ..http.proxy import HttpProxyBasePlugin
+
+import re
+
+logger = logging.getLogger(__name__)
+
+
+class FilterByURLRegexPlugin(HttpProxyBasePlugin):
+    """
+        Drop traffic by inspecting request URL,
+        checking against a list of regular expressions,
+        then returning a HTTP status code.
+    """
+
+    FILTER_LIST = [
+        {
+            'regex': b'tpc.googlesyndication.com/simgad/.*',
+            'status_code': httpStatusCodes.NOT_FOUND,
+            'notes': 'Google image ads',
+        },
+        {
+            'regex': b'tpc.googlesyndication.com/sadbundle/.*',
+            'status_code': httpStatusCodes.NOT_FOUND,
+            'notes': 'Google animated ad bundles',
+        },
+        {
+            'regex': b'pagead\\d+.googlesyndication.com/.*',
+            'status_code': httpStatusCodes.NOT_FOUND,
+            'notes': 'Google tracking',
+        },
+        {
+            'regex': b'(www){0,1}.google-analytics.com/r/collect\\?.*',
+            'status_code': httpStatusCodes.NOT_FOUND,
+            'notes': 'Google tracking',
+        },
+        {
+            'regex': b'(www){0,1}.facebook.com/tr/.*',
+            'status_code': httpStatusCodes.NOT_FOUND,
+            'notes': 'Facebook tracking',
+        },
+        {
+            'regex': b'tpc.googlesyndication.com/daca_images/simgad/.*',
+            'status_code': httpStatusCodes.NOT_FOUND,
+            'notes': 'Google image ads',
+        },
+        {
+            'regex': b'.*.2mdn.net/videoplayback/.*',
+            'status_code': httpStatusCodes.NOT_FOUND,
+            'notes': 'Twitch.tv video ads',
+        },
+        {
+            'regex': b'(www.){0,1}google.com(.*)/pagead/.*',
+            'status_code': httpStatusCodes.NOT_FOUND,
+            'notes': 'Google ads',
+        },
+    ]
+
+    def before_upstream_connection(
+            self, request: HttpParser) -> Optional[HttpParser]:
+        return request
+
+    def handle_client_request(
+            self, request: HttpParser) -> Optional[HttpParser]:
+
+        # determine host
+        request_host = None
+        if request.host:
+            request_host = request.host
+        else:
+            if b'host' in request.headers:
+                request_host = request.header(b'host')
+
+        if not request_host:
+            logger.error("Cannot determine host")
+            return request
+
+        # build URL
+        url = b'%s%s' % (
+            request_host,
+            request.path,
+        )
+
+        # check URL against list
+        rule_number = 1
+        for blocked_entry in self.FILTER_LIST:
+
+            # if regex matches on URL
+            if re.search(blocked_entry['regex'], url):
```

Comment:
@mikenye No worries, these `mypy` fixes can be notorious sometimes. Looking at the errors here https://github.com/abhinavsingh/proxy.py/pull/397/checks?check_run_id=862159685

```
Run flake8 --ignore=W504 --max-line-length=127 --max-complexity=19 proxy/ tests/ setup.py
proxy/plugin/filter_by_url_regex.py:106: error: No overload variant of "search" matches argument types "object", "bytes"
proxy/plugin/filter_by_url_regex.py:106: note: Possible overload variants:
proxy/plugin/filter_by_url_regex.py:106: note:     def [AnyStr in (str, bytes)] search(pattern: AnyStr, string: AnyStr, flags: Union[int, RegexFlag] = ...) -> Optional[Match[AnyStr]]
proxy/plugin/filter_by_url_regex.py:106: note:     def [AnyStr in (str, bytes)] search(pattern: Pattern[AnyStr], string: AnyStr, flags: Union[int, RegexFlag] = ...) -> Optional[Match[AnyStr]]
proxy/plugin/filter_by_url_regex.py:110: error: Incompatible types in string interpolation (expression has type "object", placeholder has type "Union[int, float, SupportsInt]")
proxy/plugin/filter_by_url_regex.py:119: error: Argument "status_code" to "HttpRequestRejected" has incompatible type "object"; expected "Optional[int]"
Found 3 errors in 1 file (checked 114 source files)
```

I think issue here is with `bytes` type arguments being passed to `re.search`.  Can to try to pass `str` type and see if that helps?

Reply:
Gave it a quick try locally, check https://github.com/abhinavsingh/proxy.py/commit/bc934e77436fa02305dd4354a4259398f9504598

Had to also add typing information for our `FILTER_LIST`, otherwise `mypy` will infer everything as `object` leading into `Argument "status_code" to "HttpRequestRejected" has incompatible type "object"; expected "Optional[int]"`

Reply:
Thanks so much for the help. This is my first time working with typed python - I need to do some more reading...